### ISSUE TITLE
Add visitor to remove bias from layers that are inputs to `bn_` layers

### ISSUE THREAD

**Issue by arrufat** [davisking/dlib#2155]
The third introduction to deep neural networks ([dnn_introduction3_ex.cpp](../tree/e9c56fb21aa79d7353a6152aaeb7d8df52adf32f/examples/dnn_introduction3_ex.cpp)) shows how one can define a simple visitor to modify some networks parameters.

The [dnn_dcgan_ex.cpp](../tree/5a715fe24d588469f4a4c0ab2e9d35a6afcef131/examples/dnn_dcgan_train_ex.cpp) shows an example of a visitor that removes bias learning:
https://github.com/davisking/dlib/blob/5a715fe24d588469f4a4c0ab2e9d35a6afcef131/examples/dnn_dcgan_train_ex.cpp#L40-L55

However, as you might now, removing the bias learning from a layer makes a lot of sense if that layer itself is the input of a batch normalization layer, because the batch normalization will remove the mean, anyway.

I have tried to implement a visitor that will do such a thing without success, since, whenever I visit the network layers and I find a `bn_` layer at layer `i`, I need to disable the bias learning for the layer `i+1`. But I can't do that because `i` is not a `constexpr` in the following case:

``` cpp
namespace impl
{
    template <typename net_type>
    class visitor_bn_no_bias_prev
    {
    public:

        visitor_bn_no_bias_prev(net_type& net_) : net(net_) {}

        template <typename T>
        void set_no_bias(size_t , T&) const
        {
            // ignore other layer detail types
        }

        template <layer_mode mode>
        void set_no_bias(size_t i, bn_<mode>& )
        {
            layer<i>(net).layer_details().set_bias_learning_rate_multiplier(0);
            layer<i>(net).layer_details().set_bias_weight_decay_multiplier(0);
        }

        template<typename input_layer_type>
        void operator()(size_t , input_layer_type& ) const
        {
            // ignore other layers
        }

        template <typename T, typename U, typename E>
        void operator()(size_t i, add_layer<T,U,E>& l)
        {
            set_no_bias(i + 1, l.layer_details());
        }

    private:

        net_type& net;
    };
}


template <typename net_type>
void set_all_bn_prev_no_bias (
    net_type& net
)
{
    visit_layers(net, impl::visitor_bn_no_bias_prev<net_type>(net));
}
```

I think I should look at how `visit_layers` and is implemented and specially at `impl::vl_loop`, right? Any advice?
Maybe I should make recursive templated function that calls itself with `i+1`, but I am not very familiar with this stuff (yet).

If we manage to make it work, it would be pretty neat, since, when using other toolkits, I have to do it manually at each layer followed by a batch normalization layer.

**Comment by arrufat** [comment]
In the end it was way easier than I expected, I just needed to remove the `.layer_details()` from the `operator()`, and I could just call `subnet()`. Did I ever tell you how much I like dlib for neural networks?
